Update HttpClientFactory docs (#31629)
dotnet/docs: docs/core/extensions/httpclient-factory.md

With modern application development principles driving best practices, the <xref:System.Net.Http.IHttpClientFactory> serves as a factory abstraction that can create `HttpClient` instances with custom configurations. <xref:System.Net.Http.IHttpClientFactory> was introduced in .NET Core 2.1. Common HTTP-based .NET workloads can take advantage of resilient and transient-fault-handling third-party middleware with ease.

> [!NOTE]
> If your app requires cookies, it might be better to avoid using <xref:System.Net.Http.IHttpClientFactory> in your app. For alternative ways of managing clients, see [Guidelines for using HTTP clients](../../fundamentals/networking/http/httpclient-guidelines.md).

> [!IMPORTANT]

> > Lifetime management of `HttpClient`s created by `IHttpClientFactory` is completely different from ones created manually. The strategies are to use either **short-lived** clients created by `IHttpClientFactory` or **long-lived** clients with `PooledConnectionLifetime` set up. For more information, see [`HttpClient` lifetime management](#httpclient-lifetime-management) section and [Guidelines for using HTTP clients](../../fundamentals/networking/http/httpclient-guidelines.md).

Reviewer: ```suggestion
> Lifetime management of `HttpClient`s created by `IHttpClientFactory` is completely different from ones created manually. The strategies are to use either **short-lived** clients created by `IHttpClientFactory` or **long-lived** clients with `PooledConnectionLifetime` set up. For more information, see [HttpClient lifetime management](#httpclient-lifetime-management) section and [Guidelines for using HTTP clients](../../fundamentals/networking/http/httpclient-guidelines.md).
```

1. Create an instance of `HttpClient`.
1. Create an instance of `JokeService`, passing in the instance of `HttpClient` to its constructor.

> [!IMPORTANT]

> > Using Typed clients in singleton services can be dangerous. For more information, see [Using Typed clients in singleton services](#using-typed-clients-in-singleton-services) section.

Reviewer: ```suggestion
> Using typed clients in singleton services can be dangerous. For more information, see [Using Typed clients in singleton services](#using-typed-clients-in-singleton-services) section.
```

```

> [!IMPORTANT]
> `HttpClient` instances created by `IHttpClientFactory` are intended to be **short-lived**.

> > 1. Recycling and recreating `HttpMessageHandler`s when their lifetime expires is essential for `IHttpClientFactory` to ensure the handlers react to DNS changes. `HttpClient` is tied to a specific handler instance upon it's creation, so new `HttpClient` instances should be requested in a timely manner to ensure the client will get the updated handler.

Reviewer: ```suggestion
> 1. Recycling and recreating `HttpMessageHandler`'s when their lifetime expires is essential for `IHttpClientFactory` to ensure the handlers react to DNS changes. `HttpClient` is tied to a specific handler instance upon its creation, so new `HttpClient` instances should be requested in a timely manner to ensure the client will get the updated handler.
>
```

> > 2. Disposing such `HttpClient`s **created by the factory** will not lead to socket exhaustion, as it's disposal **will not** trigger `HttpMessageHandler`'s disposal. `IHttpClientFactory` tracks and disposes resources used to create `HttpClient` instances, specifically the `HttpMessageHandler`, as soon it's lifetime expires and there's no `HttpClient`s using it anymore.

Reviewer: ```suggestion
> 1. Disposing of such `HttpClient`s **created by the factory** will not lead to socket exhaustion, as its disposal **will not** trigger `HttpMessageHandler`'s disposal. `IHttpClientFactory` tracks and disposes of resources used to create `HttpClient` instances, specifically the `HttpMessageHandler`, as soon its lifetime expires and there's no `HttpClient` using it anymore.
```

However, `SocketsHttpHandler` and `IHttpClientFactory` can be used together, to benefit from configurability on both low level (e.g. using `LocalCertificateSelectionCallback` for dynamic certificate selection) and high level (e.g. leveraging DI integration and several client configurations).

To do that, you need to
1. Specify `SocketsHttpHandler` as `PrimaryHandler` and set up it's `PooledConnectionLifetime` (e.g. to a value that was previously in `HandlerLifetime`)
2. As `SocketsHttpHandler` will handle connection pooling and recycling, then `IHttpClientFactory`-level handler recycling is not needed anymore, so it could be disabled by setting `HandlerLifetime` to `Timeout.InfiniteTimeSpan`.

```csharp
services.AddHttpClient(name)
    .ConfigurePrimaryHttpMessageHandler(() =>
    {
        return new SocketsHttpHandler()
        {

>             PooledConnectionLifetime = TimeSpan.FromMinutes(handlerLifetimeInMinutes)

Author: Nowhere 😅 kind of an example of thing that a user could have around. The code was mostly from https://github.com/dotnet/runtime/discussions/76064. I didn't want to use a specific constant there, but if you think it will be better, I can change it to `TimeSpan.FromMinutes(2)` for example.

>             PooledConnectionLifetime = TimeSpan.FromMinutes(handlerLifetimeInMinutes)

Reviewer: Yes, I think that would make a lot more sense. Thanks

        };
    })
    .SetHandlerLifetime(Timeout.InfiniteTimeSpan); // Disable rotation, as it is handled by PooledConnectionLifetime
```

## Using Typed clients in singleton services

> When Named client approach is used, `IHttpClientFactory` is injected into services, and `HttpClient`s are created by calling <xref:System.Net.Http.IHttpClientFactory.CreateClient%2A> every time an `HttpClient` is needed.

Reviewer: ```suggestion
When using the _named client_ approach, `IHttpClientFactory` is injected into services, and `HttpClient`s are created by calling <xref:System.Net.Http.IHttpClientFactory.CreateClient%2A> every time an `HttpClient` is needed.
```

> However, with Typed client approach, Typed clients are transient objects usually injected into services. That may cause a problem, because a Typed client can be injected into a singleton service.

Reviewer: ```suggestion
However, with the _typed client_ approach, typed clients are transient objects usually injected into services. That may cause a problem because a typed client can be injected into a singleton service.
```

> [!IMPORTANT]
> Typed clients are expected to be **short-lived** in the same sense as `HttpClient`s created by `IHttpClientFactory`, see [`HttpClient` lifetime management](#httpclient-lifetime-management) section. As soon as a Typed client instance is created, `IHttpClientFactory` has no control over it. If a Typed client instance is captured in a singleton, it may prevent it from reacting to DNS changes, defeating one of the purposes of `IHttpClientFactory`.

If you need to use `HttpClient`s in a singleton service, consider the following options:
1. Use Named client approach instead, injecting `IHttpClientFactory` in the singleton service and recreating `HttpClient`s when nesessary.

> 2. If you require a Typed client approach, use `SocketsHttpHandler` with configured `PooledConnectionLifetime` as a primary handler. For more information on using `SocketsHttpHandler` with `IHttpClientFactory`, see [Using HttpClientFactory together with SocketsHttpHandler](#using-httpclientfactory-together-with-socketshttphandler) section.

Reviewer: ```suggestion
1. If you require a typed client approach, use `SocketsHttpHandler` with configured `PooledConnectionLifetime` as a primary handler. For more information on using `SocketsHttpHandler` with `IHttpClientFactory`, see [Using HttpClientFactory together with SocketsHttpHandler](#using-httpclientfactory-together-with-socketshttphandler) section.
```

| <xref:Microsoft.Extensions.DependencyInjection.HttpClientBuilderExtensions.RedactLoggedHeaders%2A> | Sets the collection of HTTP header names for which values should be redacted before logging. |
| <xref:Microsoft.Extensions.DependencyInjection.HttpClientBuilderExtensions.SetHandlerLifetime%2A> | Sets the length of time that a `HttpMessageHandler` instance can be reused. Each named client can have its own configured handler lifetime value. |

## Using HttpClientFactory together with SocketsHttpHandler

> Since .NET Core 2.1, `SocketsHttpHandler` implementation of `HttpMessageHandler` was added, that allows configuring `PooledConnectionLifetime`. This setting is used to ensure that the handler will react to DNS changes, so using `SocketsHttpHandler` is considered to be an alternative to using `IHttpClientFactory`, see [Guidelines for using HTTP clients](../../fundamentals/networking/http/httpclient-guidelines.md).

Reviewer 2: ```suggestion
The `SocketsHttpHandler` implementation of `HttpMessageHandler` was added in .NET Core 2.1, which allows `PooledConnectionLifetime` to be configured. This setting is used to ensure that the handler reacts to DNS changes, so using `SocketsHttpHandler` is an alternative to using `IHttpClientFactory`. For more information, see [Guidelines for using HTTP clients](../../fundamentals/networking/http/httpclient-guidelines.md).
```

> However, `SocketsHttpHandler` and `IHttpClientFactory` can be used together, to benefit from configurability on both low level (e.g. using `LocalCertificateSelectionCallback` for dynamic certificate selection) and high level (e.g. leveraging DI integration and several client configurations).

Reviewer 2: ```suggestion
However, `SocketsHttpHandler` and `IHttpClientFactory` can be used together to improve configurability. By using both of these APIs, you benefit from configurability on both a low level (for example, using `LocalCertificateSelectionCallback` for dynamic certificate selection) and a high level (for example, leveraging DI integration and several client configurations).
```

> To do that, you need to

Reviewer 2: ```suggestion
To use both APIs:

```

> 1. Specify `SocketsHttpHandler` as `PrimaryHandler` and set up it's `PooledConnectionLifetime` (e.g. to a value that was previously in `HandlerLifetime`)

Reviewer 2: ```suggestion
1. Specify `SocketsHttpHandler` as `PrimaryHandler` and set its `PooledConnectionLifetime` (for example, to a value that was previously in `HandlerLifetime`).
```

> 2. As `SocketsHttpHandler` will handle connection pooling and recycling, then `IHttpClientFactory`-level handler recycling is not needed anymore, so it could be disabled by setting `HandlerLifetime` to `Timeout.InfiniteTimeSpan`.

Reviewer 2: ```suggestion
2. As `SocketsHttpHandler` will handle connection pooling and recycling, then handler recycling at the `IHttpClientFactory` level is not needed anymore. You can disable it by setting `HandlerLifetime` to `Timeout.InfiniteTimeSpan`.
```

```csharp
services.AddHttpClient(name)
    .ConfigurePrimaryHttpMessageHandler(() =>
    {
        return new SocketsHttpHandler()
        {
            PooledConnectionLifetime = TimeSpan.FromMinutes(handlerLifetimeInMinutes)
        };
    })
    .SetHandlerLifetime(Timeout.InfiniteTimeSpan); // Disable rotation, as it is handled by PooledConnectionLifetime
```

> ## Using Typed clients in singleton services

Reviewer 2: ```suggestion
## Use typed clients in singleton services
```

> When using the _named client_ approach, `IHttpClientFactory` is injected into services, and `HttpClient`s are created by calling <xref:System.Net.Http.IHttpClientFactory.CreateClient%2A> every time an `HttpClient` is needed.

Reviewer 2: ```suggestion
When using the _named client_ approach, `IHttpClientFactory` is injected into services, and `HttpClient` instances are created by calling <xref:System.Net.Http.IHttpClientFactory.CreateClient%2A> every time an `HttpClient` is needed.
```

However, with the _typed client_ approach, typed clients are transient objects usually injected into services. That may cause a problem because a typed client can be injected into a singleton service.

> [!IMPORTANT]

> > Typed clients are expected to be **short-lived** in the same sense as `HttpClient`s created by `IHttpClientFactory`, see [`HttpClient` lifetime management](#httpclient-lifetime-management) section. As soon as a Typed client instance is created, `IHttpClientFactory` has no control over it. If a Typed client instance is captured in a singleton, it may prevent it from reacting to DNS changes, defeating one of the purposes of `IHttpClientFactory`.

Reviewer 2: ```suggestion
> Typed clients are expected to be **short-lived** in the same sense as `HttpClient` instances created by `IHttpClientFactory` (for more information, see [`HttpClient` lifetime management](#httpclient-lifetime-management)). As soon as a typed client instance is created, `IHttpClientFactory` has no control over it. If a typed client instance is captured in a singleton, it may prevent it from reacting to DNS changes, defeating one of the purposes of `IHttpClientFactory`.
```

> If you need to use `HttpClient`s in a singleton service, consider the following options:

Reviewer 2: ```suggestion
If you need to use `HttpClient` instances in a singleton service, consider the following options:

```

> 1. Use Named client approach instead, injecting `IHttpClientFactory` in the singleton service and recreating `HttpClient`s when nesessary.

Reviewer 2: ```suggestion
- Use the named client approach instead, injecting `IHttpClientFactory` in the singleton service and recreating `HttpClient` instances when necessary.
```

> 1. If you require a typed client approach, use `SocketsHttpHandler` with configured `PooledConnectionLifetime` as a primary handler. For more information on using `SocketsHttpHandler` with `IHttpClientFactory`, see [Using HttpClientFactory together with SocketsHttpHandler](#using-httpclientfactory-together-with-socketshttphandler) section.

Reviewer 2: ```suggestion
- If you require a typed client approach, use `SocketsHttpHandler` with configured `PooledConnectionLifetime` as a primary handler. For more information on using `SocketsHttpHandler` with `IHttpClientFactory`, see the section [Using HttpClientFactory together with SocketsHttpHandler](#using-httpclientfactory-together-with-socketshttphandler).
```
